repos/system_upgrade/common/actors/systemd/transitionsystemdservicesstates/libraries/transitionsystemdservicesstates.py

```diff
@@ -0,0 +1,211 @@
+
+    :param service_name: Then name of the service
+    :param desired_state: The state the service should set to
+    :param state_target: State on the target system
+    :param tasks: The tasks to append the task to
+    """
+    if desired_state == state_target:
+        return
+
+    if desired_state == "enabled":
+        tasks.to_enable.append(service_name)
+    if desired_state == "disabled":
+        tasks.to_disable.append(service_name)
+
+
+def _get_service_preset(service_name, presets):
+    preset = presets.get(service_name)
+    if not preset:
+        # shouldn't really happen as there is usually a `disable *` glob as
+        # the last statement in the presets
+        api.current_logger().debug(
+            'No presets found for service "{}", assuming "disable"'.format(service_name)
+        )
+        return "disable"
+    return preset
+
+
+def _filter_services(services_source, services_target):
+    """
+    Filter out irrelevant services
+    """
+    filtered = []
+    for service in services_target:
+        if service.state not in ("enabled", "disabled", "enabled-runtime"):
+            # Enabling/disabling of services is only relevant to these states
+            continue
+
+        state_source = services_source.get(service.name)
+        if not state_source:
+            # The service doesn't exist on the source system
+            continue
+
+        if state_source == "masked-runtime":
+            # TODO(mmatuska): It's not possible to get the persistent
+            # (non-runtime) state of a service with `systemctl`. One solution
+            # might be to check symlinks
+            api.current_logger().debug(
+                'Skipping service in "masked-runtime" state: {}'.format(service.name)
+            )
+            continue
+
+        filtered.append(service)
+
+    return filtered
+
+
+def _get_required_tasks(services_target, desired_states):
+    """
+    Get the required tasks to set the services on the target system to their desired state
+
+    :return: The tasks required to be executed
+    :rtype: SystemdServicesTasks
+    """
+    tasks = SystemdServicesTasks()
+
+    for service in services_target:
+        desired_state = desired_states[service.name]
+        _get_service_task(service.name, desired_state, service.state, tasks)
+
+    return tasks
+
+
+def _report_kept_enabled(tasks):
+    summary = (
+        "Systemd services which were enabled on the system before the upgrade"
+        " were kept enabled after the upgrade. "
+    )
+    if tasks:
+        summary += (
+            "The following services were originally disabled on the upgraded system"
+            " and Leapp attempted to enable them:{}{}"
+        ).format(FMT_LIST_SEPARATOR, FMT_LIST_SEPARATOR.join(sorted(tasks.to_enable)))
+        # TODO(mmatuska): When post-upgrade reports are implemented in
+        # `setsystemdservicesstates actor, add a note here to check the reports
+        # if the enabling failed
```

Comment:
@matejmatuska regarding the current implementation, I expect this will be always true. Add please unit tests that will check the correct summary is picked based on the input.

Reply:
Seems this is the only problem I found by the review of the code. It seems to me good otherwise. I will get to the testing during the week.

Reply:
> @matejmatuska regarding the current implementation, I expect this will be always true.

I suspect that there might be a system which doesn't meet this condition, so I made this conditional. Also I think it's more future-proof.

I will add the test.

Reply:
I see, it make sense. Thanks for the explanation.